### DISCUSSION TITLE
How to update a field in Settings model in afterSave()

### DISCUSSION THREAD

**Discussion by AIC-BV** [wintercms/winter#397] [Q&A] [answered]
Hi,

I am trying to update a field using afterSave() in my Settings model.
But I keep running into walls.

I can read my field like this:
`$this->value['estimated_delivery_date'];`.

Updating my field like this:
`$this->value['estimated_delivery_date'] = $estimated_delivery_date; `
--> returns an error: "Indirect modification of overloaded property has no effect"

Updating my field like this:
`Settings::set('estimated_delivery_date', $estimated_delivery_date);`
--> returns an error: "internal server error 500"
--> BUT when refreshing the page, the value succesfuly changed.

Same result as above with this (error 500):
```
$settings = Settings::instance();
$settings->estimated_delivery_date = $estimated_delivery_date;
$settings->save();
```

I created a hacky workaround like this:
```
$my_value = json_decode(Db::table('system_settings')->where('item', 'delivery_settings')->value('value'));
$my_value->estimated_delivery_date = $estimated_delivery_date;
Db::table('system_settings')->where('item', 'delivery_settings')->update(['value' => json_encode($my_value)]);
return redirect()->refresh();
```
--> Works, but the page doesn't refresh

Does anyone have an idea or solution?
Thanks!

**Comment by LukeTowers** [comment, marked as answer]
Why are you doing it in afterSave() instead of beforeSave()?

**Reply by LukeTowers** [reply]
saving a record in afterSave will create an infinite loop

**Reply by AIC-BV** [reply]
Getting a 500 in beforeSave() too

I thought afterSave() was the correct method because it first saves the changes to the model, and afterSave() I can pick up those changes and use them to calculate a new delivery date

**Reply by LukeTowers** [reply]
If it's a calculated property why not use an accessor instead of persisting the value in the DB in the first place?

**Reply by AIC-BV** [reply]
Hmm..

I think I want it like this because the delivery date only needs to be recalculated when:
- every day at midnight
- settings are changed (for example; deliveries are no longer possible on saturdays and sundays)
- a new order comes in (because the day might be fully booked, we have to deliver it by truck)

This sounds more performant to me than recalculating the delivery date on every client visiting every webpage. (date needs to be visible to visitors)
Having the data stored in a field, gives easy access to printing it on the front- and back-end in a readOnly field.

**Reply by LukeTowers** [reply]
You could also achieve the same results with using the Cache in the accessor I believe

**Reply by AIC-BV** [reply]
Cache looks promising! I'll keep you up to date :) thanks!

**Reply by AIC-BV** [reply]
I don't really understand what the big difference is, but changing
`Settings::set('estimated_delivery_date', $estimated_delivery_date);` in the afterSave() into
`Cache::forever('estimated_delivery_date', $estimated_delivery_date);` did the trick.

```
public function afterSave()
{
    $delivery = new Delivery();
    $delivery->calculateEstimatedDeliveryDate();
    return Redirect::refresh();
}
```
(The calculation itself is in a class Delivery which has the Cache::forever line)

The value gets saved somewhere in the cache, I'm no longer getting the internal server errors and I can read it in the frontend with this:
```
public function boot()
{
    Event::listen('cms.page.beforeDisplay', function($controller, $url, $page) {
        $controller->vars['estimated_delivery_date'] = Cache::get('estimated_delivery_date');
    });
}
==
{{ estimated_delivery_date | date('d-m-Y') }}
```
and in the backend I can read it in a partial like this: `<?= Cache::get('estimated_delivery_date')->format('d-m-Y') ?>`

To update the value daily I will register a schedule that updates the Cached value at midnight.
```
public function registerSchedule($schedule)
    {
        $schedule->call(function () {
            $delivery = new Delivery();
            $delivery->calculateEstimatedDeliveryDate();
        })->daily();
    }
```

The only issue I'm facing is that the page doesn't refresh afterSave() to display the recalculated value that was updated in the Cache (`return Redirect::refresh();` is not working here?).
I don't think I used accessors/mutators. Do you think this is a fine solution?

**Reply by LukeTowers** [reply]
What I would recommend instead is provide a `getEstimatedDeliveryDate()` method on your Delivery service and call that in your frontend instead:
```php
public function getEstimatedDeliveryDate()
{
    return Cache::remember('estimated_delivery_date', now()->addDays(1), function () {
        return $this->calculateEstimatedDeliveryDate();
    });
}
```
and
```php
public function boot()
{
    Event::listen('cms.page.beforeDisplay', function($controller, $url, $page) {
        $controller->vars['estimated_delivery_date'] = (new Delivery())->getEstimatedDeliveryDate();
    });
}
==
{{ estimated_delivery_date | date('d-m-Y') }}
```
This will remove the need for a separate scheduled action to take place since once per day the cache will be automatically invalidated and regenerate itself on demand and it will remove the need for interacting with the Cache directly in your view which is good because your current code will fail if there's nothing in the cache yet while the above example will work just fine even if you just ran `cache:clear` for whatever reason.

**Reply by AIC-BV** [reply]
Brilliant! Works like a charm. Thank you!
